libs/beacon/src/beacon_complaints_manager.cpp

```diff
@@ -0,0 +1,319 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "beacon/beacon_complaints_manager.hpp"
+
+#include <cstdint>
+#include <mutex>
+
+namespace fetch {
+namespace beacon {
+
+constexpr char const *LOGGING_NAME = "DKGComplaints";
+
+void ComplaintsManager::ResetCabinet(MuddleAddress const &address, uint32_t threshold)
+{
+  FETCH_LOCK(mutex_);
+  threshold_ = threshold;
+  address_   = address;
+  finished_  = false;
+  complaints_counter_.clear();
+  complaints_from_.clear();
+  complaints_.clear();
+  complaints_received_.clear();
+}
+
+void ComplaintsManager::AddComplaintAgainst(MuddleAddress const &complaint_address)
+{
+  FETCH_LOCK(mutex_);
+  complaints_counter_[complaint_address].insert(address_);
+}
+
+void ComplaintsManager::AddComplaintsFrom(ComplaintsMessage const &msg,
+                                          MuddleAddress const &    from_id)
+{
+  FETCH_LOCK(mutex_);
+  // Check if we have received a complaints message from this node before and if not log that we
+  // received a complaint message
+  if (complaints_received_.find(from_id) == complaints_received_.end())
+  {
+    complaints_received_.insert(from_id);
+  }
+  else
+  {
+    FETCH_LOG_WARN(LOGGING_NAME, "Duplicate complaints received. Discarding.");
+    return;
+  }
+
+  for (auto const &bad_node : msg.complaints())
+  {
+    complaints_counter_[bad_node].insert(from_id);
+    // If a node receives complaint against itself then store in complaints from
+    // for answering later
+    if (bad_node == address_)
+    {
+      complaints_from_.insert(from_id);
+    }
+  }
+}
+
+void ComplaintsManager::Finish(std::set<Identity> const &cabinet)
+{
+  FETCH_LOCK(mutex_);
+  if (!finished_)
```

Comment:
I thought that the concept of 'finishing' was going away as you only fail on qual?

Reply:
The finish is not about failing. It is about telling the complaints manager we have ended the wait for complaints messages, which is important as it then adds to complaints all nodes it did not get a message from.